Convert JS to TS (commits 7-13 of PR435) - attempt#2

webwarrior-ws opened the pull request:
Supersedes #557. Added a commit that should fix bug in disputes.

Catrya reviewed (commented):
Hi @webwarrior-ws @Mersho 
When the admin goes to take a dispute this error appears in the logs:
```
[2024-09-04T10:09:59.680-04:00] error: Community was not found in DB Error: Community was not found in DB
    at /home/catry/bot/bot/validations.js:208:27
    at step (/home/catry/bot/bot/validations.js:56:23)
    at Object.next (/home/catry/bot/bot/validations.js:37:53)
    at fulfilled (/home/catry/bot/bot/validations.js:28:58)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Catrya reviewed (commented):
It also happened to me that the wizard_add_invoice_init message was not being sent in any order even though it entered the wizard because any command or invoice that I sent to it would delete it. However, I stopped it and started it again and it worked.

Catrya reviewed (commented):
When sending an invoice to receive, it always returns this error to the buyer: `Invoice expiry is too short`

webwarrior-ws commented:
The bugs should be fixed now, but I haven't tested.

Catrya commented:
> Hola@webwarrior-ws @Mersho Cuando el administrador va a tomar una disputa aparece este error en los registros:
> 
> ```
> [2024-09-04T10:09:59.680-04:00] error: Community was not found in DB Error: Community was not found in DB
>     at /home/catry/bot/bot/validations.js:208:27
>     at step (/home/catry/bot/bot/validations.js:56:23)
>     at Object.next (/home/catry/bot/bot/validations.js:37:53)
>     at fulfilled (/home/catry/bot/bot/validations.js:28:58)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> ```

This error has already been solved

Catrya reviewed (commented):
> Al enviar una factura para recibir, siempre devuelve este error al comprador:`Invoice expiry is too short`

@webwarrior-ws @Mersho  That error is still there, it always says `Invoice expiry is too short` but if I change to or any other branch and put the same invoice it accepts it without problem

webwarrior-ws commented:
> > Al enviar una factura para recibir, siempre devuelve este error al comprador:`Invoice expiry is too short`
> 
> @webwarrior-ws @Mersho That error is still there, it always says `Invoice expiry is too short` but if I change to or any other branch and put the same invoice it accepts it without problem

I tried with my test setup and could not reproduce this bug.

webwarrior-ws commented:
Added some debug logging. Maybe that will help with identifying the problem.

Catrya commented:
> Added some debug logging. Maybe that will help with identifying the problem.

Well, the bot did not show the `wizard_add_invoice_init` message and this error appeared: 
`[2024-09-09T09:49:10.191-04:00] error: All promises were rejected AggregateError: All promises were rejected`

But I knew it was 500 sats so I put the invoice and this appears: 
```
Date(invoice.expires_at) = Mon Sep 09 2024 10:49:28 GMT-0400 (hora de verano de Cuba)
latestDate = Mon Sep 09 2024 10:49:47 GMT-0400 (hora de verano de Cuba)
INVOICE_EXPIRATION_WINDOW = 3600000
```

webwarrior-ws commented:
I don't see any bug in lnp2pBot code here. Invoice expires less than `INVOICE_EXPIRATION_WINDOW` (1 hour) from time of running, and that results in error.
I don't know why `main` branch doesn't result in error in the same scenario. Maybe this code never worked as intended? I noticed that in original code, Date is compared to string.

knocte commented:
> Maybe this code never worked as intended? I noticed that in original code, Date is compared to string.

Looks like we need @grunch's input here?

webwarrior-ws commented:
@grunch I have tried to reproduce the bug, but I can't take a dispute - when I click on "Take dispute" button I get a message "You have been banned!". As I only have 2 telegram accounts, solver user is one of the participants, maybe that's the issue?

webwarrior-ws commented:
Me and @knocte managed to take and resolve dispute in community. That error looks like contents of some variable is not escaped properly. I've seen a commit in `main` that adds escaping. I rebased this branch. Maybe this will also solve the bug. If not, I added a commit that dumps contents of the problematic message into console.

grunch commented:
> @grunch I have tried to reproduce the bug, but I can't take a dispute - when I click on "Take dispute" button I get a message "You have been banned!". As I only have 2 telegram accounts, solver user is one of the participants, maybe that's the issue?

Yes, we ban users after being involved on n disputes, I think that restriction is not useful and we should remove it in the future, but to continue just remove the ban on the document for that user, change field `banned` to false, then try again

knocte commented:
>  change field banned to false, then try again

Francisco, that is old info. After webwarrior wrote that message, I assisted him in his testing, being the 3rd user in his setup (the solver, in particular), and he couldn't reproduce the issue. See https://github.com/lnp2pBot/bot/pull/585#issuecomment-2346186472 , and test again with the new logs added please.

grunch commented:
> @grunch I have tried to reproduce the bug, but I can't take a dispute - when I click on "Take dispute" button I get a message "You have been banned!". As I only have 2 telegram accounts, solver user is one of the participants, maybe that's the issue?



Hi @webwarrior-ws sorry for the delay, are you able to run the bot in this branch? I'm testing and I'm getting this errors

```
[2024-09-16T17:22:26.672-03:00] error: ResubcribeInvoice catch: 503,UnexpectedGetInvoicesError,[object Object] 
[2024-09-16T17:23:00.027-03:00] error: 503,FailedToConnectToDaemon 
[2024-09-16T17:23:00.028-03:00] error: node info catch error: TypeError: Cannot read properties of undefined (reading 'is_synced_to_chain') 
[2024-09-16T17:24:00.022-03:00] error: 503,FailedToConnectToDaemon 
[2024-09-16T17:24:00.022-03:00] error: node info catch error: TypeError: Cannot read properties of undefined (reading 'is_synced_to_chain') 
```

I that we are having conflics, can you solve them and we can try again?

webwarrior-ws commented:
@grunch I've rebased PR branch and there are no merge conflicts now. I've also tested it with standard sell/buy procedure and got no 503 errors that you have mentioned.

grunch reviewed (changes requested):
Thank you for solving the conflicts, I just tested it again and I found a new bug, here the log:
```
[2024-09-24T14:56:16.092-03:00] error: 400: Bad Request: can't parse entities: Character '.' is reserved and must be escaped with the preceding '\' Error: 400: Bad Request: can't parse entities: Character '.' is reserved and must be escaped with the preceding '\'
    at Telegram.callApi (/home/grunch/dev/bot-warrior/node_modules/telegraf/lib/core/network/client.js:315:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async exports.disputeData (/home/grunch/dev/bot-warrior/bot/modules/dispute/messages.js:114:5)
```

I dig in a bit and I found that you can't use object properties inside locales, so I changed, here the patch, can you apply this on the other languages? thanks

```patch
diff --git a/bot/modules/dispute/messages.js b/bot/modules/dispute/messages.js
index a887012..d8bc775 100644
--- a/bot/modules/dispute/messages.js
+++ b/bot/modules/dispute/messages.js
@@ -1,4 +1,8 @@
-const { getDisputeChannel, getDetailedOrder, sanitizeMD } = require('../../../util');
+const {
+  getDisputeChannel,
+  getDetailedOrder,
+  sanitizeMD,
+} = require('../../../util');
 const { logger } = require('../../../logger');
 
 exports.beginDispute = async (ctx, initiator, order, buyer, seller) => {
@@ -91,11 +95,11 @@ exports.disputeData = async (
     // Fix Issue 543: Escape underscores in usernames
     const escapedInitiatorUsername = sanitizeMD(initiatorUser.username);
     const escapedCounterPartyUsername = sanitizeMD(counterPartyUser.username);
-    
+
     const message = ctx.i18n.t('dispute_started_channel', {
-      initiatorUser: { ...initiatorUser, username: escapedInitiatorUsername },
+      initiatorUser: escapedInitiatorUsername,
       initiatorTgId: initiatorUser.tg_id,
-      counterPartyUser: { ...counterPartyUser, username: escapedCounterPartyUsername },
+      counterPartyUser: escapedCounterPartyUsername,
       counterPartyUserTgId: counterPartyUser.tg_id,
       buyer,
       seller,
@@ -107,11 +111,9 @@ exports.disputeData = async (
       buyerToken: order.buyer_dispute_token,
     });
     console.log(`Contens of message:\n${message}`);
-    await ctx.telegram.sendMessage(
-      solver.tg_id,
-      message,
-      { parse_mode: 'MarkdownV2' }
-    );
+    await ctx.telegram.sendMessage(solver.tg_id, message, {
+      parse_mode: 'MarkdownV2',
+    });
     // message to both parties letting them know the dispute
     // has been taken by a solver
     await ctx.telegram.sendMessage(
diff --git a/locales/es.yaml b/locales/es.yaml
index 9baf03f..26cbc22 100644
--- a/locales/es.yaml
+++ b/locales/es.yaml
@@ -202,16 +202,16 @@ order_detail: |
 seller: vendedor
 buyer: comprador
 dispute_started_channel: |
-  El ${type} @${initiatorUser.username} TG ID: ${initiatorTgId}
-  ha iniciado una disputa con @${counterPartyUser.username} TG ID: ${counterPartyUserTgId} en la orden:
+  El ${type} @${initiatorUser} TG ID: ${initiatorTgId}
+  ha iniciado una disputa con @${counterPartyUser} TG ID: ${counterPartyUserTgId} en la orden:
 
   ${detailedOrder}
 
   Seller Token: ${sellerToken}
   Buyer Token: ${buyerToken}
 
-  @${initiatorUser.username} ya tiene ${buyerDisputes} disputas
-  @${counterPartyUser.username} ya tiene ${sellerDisputes} disputas
+  @${initiatorUser} ya tiene ${buyerDisputes} disputas
+  @${counterPartyUser} ya tiene ${sellerDisputes} disputas
 you_started: '🥴 Has iniciado una disputa en tu orden con Id: ${orderId}.'
 counterpart_started: '🥴 Tu contraparte ha iniciado una disputa en tu orden con Id: ${orderId}.'
 dispute_started: '${who} Un solver te atenderá pronto, cuando él/la solver sea asignado a tu disputa el bot te dirá su username, solo él/ella podrá atenderte. Puedes escribirle directamente, pero si él/ella te contacta primero, debes pedirle que te diga cuál es el token de tu disputa, tu token es: ${token}.'
diff --git a/util/index.ts b/util/index.ts
index 0b1bc7e..902d9e4 100644
--- a/util/index.ts
+++ b/util/index.ts
@@ -368,7 +368,7 @@ const getDetailedOrder = (i18n: I18nContext, order: IOrder, buyer: UserDocument,
     takenAt = sanitizeMD(takenAt);
     const previousDisputeStatus = sanitizeMD(order.previous_dispute_status);
     const status = sanitizeMD(order.status);
-    const fee = order.fee ? Number(order.fee) : '';
+    const fee = order.fee ? sanitizeMD(Number(order.fee)) : '';
     const creator =
       order.creator_id === buyerId ? buyerUsername : sellerUsername;
     const buyerAge = buyer? getUserAge(buyer) : '';

```

webwarrior-ws commented:
@grunch applied your patch, updated locales and disputes test

grunch commented:
> @grunch applied your patch, updated locales and disputes test

great job, thanks for this contribution, as we are receiving some donations we want to give back some sats to our contributors, it's more of a symbolic gesture of gratitude 😃 

Please add as a comment a lightning invoice with amount of 200,000 sats

webwarrior-ws commented:
lnbc2m1pn0dtz0pp5ru28d55mfjw2aypjvsu7xnd5uj7363mz095slnwmca05mcequ9nscqpjsp5vuprh07ze5nk2k6rhx7zjy67axfqmsej9gmt6a7zudhsy7ej8zas9q7sqqqqqqqqqqqqqqqqqqqsqqqqqysgqdqqmqz9gxqyjw5qrzjqwryaup9lh50kkranzgcdnn2fgvx390wgj5jd07rwr3vxeje0glcll6lyxlsu4fuhqqqqqlgqqqqqeqqjqu7s52ap9967fnyvqc9nphcccsqpl3txa679zpzw2f0h6gwqhex7qwf5y2eywlmd5wyjx9e7k5ckn2luml34c304mtyuze87zy5mkfdgpmqngrw

grunch commented:
> lnbc2m1pn0dtz0pp5ru28d55mfjw2aypjvsu7xnd5uj7363mz095slnwmca05mcequ9nscqpjsp5vuprh07ze5nk2k6rhx7zjy67axfqmsej9gmt6a7zudhsy7ej8zas9q7sqqqqqqqqqqqqqqqqqqqsqqqqqysgqdqqmqz9gxqyjw5qrzjqwryaup9lh50kkranzgcdnn2fgvx390wgj5jd07rwr3vxeje0glcll6lyxlsu4fuhqqqqqlgqqqqqeqqjqu7s52ap9967fnyvqc9nphcccsqpl3txa679zpzw2f0h6gwqhex7qwf5y2eywlmd5wyjx9e7k5ckn2luml34c304mtyuze87zy5mkfdgpmqngrw

Payment hash:   1f1476d29b4c9cae90326439e34db4e4bd1d476279690fcddbc75f4de320e167
Payment status: SUCCEEDED, preimage: ed4600e227269b6d29068d1b7eb51e8bd77f9137b94e334e6bcb561b5ffcc7cf
